project_euler/problem_345/sol1.py

```diff
@@ -0,0 +1,177 @@
+    has_booked          --  List of j's flag
+    value               --  Accumulative value from parent to current node
+    max_value           --  Maximum value is stored right now by the tree
+
+    """
+
+    # Get length of a row and length of a column, respectively denoted by m and n
+    m, n = len(matrix), len(matrix[0])
+
+    # Create list of j's flag at the first time
+    if has_booked is None:
+        has_booked = [False for _ in range(m)]
+
+    # Create enhanced matrix at the first time
+    if enhanced_matrix is None:
+        # Pair value with its column index
+        enhanced_matrix = [[(row[j], j) for j in range(len(row))] for row in matrix]
+
+        # Sort each row by considering the value as comparison
+        for k in range(len(enhanced_matrix)):
+            enhanced_matrix[k] = sorted(
+                enhanced_matrix[k],
+                key=cmp_to_key(lambda a, b: 1 if a[0] < b[0] else -1),
+            )
+
+    # Base case of recursive, will happening when it reaches the leaves or last row
+    if n - 1 == i:
+        for j in range(n):
+            # Return last value
+            if not has_booked[enhanced_matrix[i][j][1]]:
+                return max(max_value, value + enhanced_matrix[i][j][0])
+
+    # Start from this line,
+    # it will run rest line codes when current computation not in the leaf
+    for j in range(n):
+        # Ensure current node don't shares same column as current selected nodes/values
+        if not has_booked[enhanced_matrix[i][j][1]]:
+            has_booked[enhanced_matrix[i][j][1]] = True
+
+            # Try to see the maximum value could be reached from current node.
+            # Take not booked columns, starts from (i+1)-th row
+            #   and each of them find the largest one and accumulate them.
+            # In the end of loop, combine current value with the accumulative value
+            #   and become upper bound.
+            possible_max_value = value + enhanced_matrix[i][j][0]
+            for k in range(len(has_booked)):
+                if not has_booked[k]:
+                    sub_max_value = float("-inf")
+                    for c in range(i + 1, m):
+                        if sub_max_value < matrix[c][k]:
+                            sub_max_value = matrix[c][k]
+                    possible_max_value += sub_max_value
+
+            # Stop evaluate next sub-tree
+            #   if there is no way to overcome current max value
+            if possible_max_value > max_value:
+                max_value = solve(
+                    matrix,
+                    enhanced_matrix,
+                    i + 1,
+                    has_booked,
+                    value + enhanced_matrix[i][j][0],
+                    max_value,
+                )
+
+            # Ensure current node is not booked anymore and others can use it
+            has_booked[enhanced_matrix[i][j][1]] = False
+
+    return max_value
+
+
+# Problem inputs
+matrix1 = [
```

Comment:
These are constants so they will be in uppercase and comes after all the imports.